ask: handle real linear inequalities

TiloRC opened the pull request:
<!-- Your title above should be a short description of what
was changed. Do not include the issue number in the title. -->
Here are some examples of what this PR enables ask to do:

```python
x, y, z = symbols("x y z", real=True)
ask(x + y > 2, (x < 0) & (y < 0)) # returns False instead of None
ask(x > z, (x > y) & (y > z))     # returns True instead of None
```

Unfortunately, I didn't have time to figure out how to handle infinite values, non-rational values, or complex variables. In the future, it might make sense to remove the `lra_satask` function and integrate the code from this pull request into `satask` once these cases are handled. Then again, one advantage of keeping it separate might be speed.

Currently, variables need to be specified to be real when they are created using the old assumptions. In other words, queries like `ask(x > 0, (x > 1) & Q.real(x))` don't currently work.

Also, if a disallowed predicate such as Q.even is part of a query, that disables inequality handling. This could be changed fairly easily. However, I wanted to avoid slowing down queries that weren't primarily about inequalities.

#### logic/algorithms/lra_theory.py

Contains the LRASolver object which is designed to work in conjunction with a SAT solver to determine if a boolean formula involving inequalities is satisfiable or not. 

#### sympy/assumptions/lra_satask.py

A function that's called in `ask` when `satask` returns None. It does some preprocessing so that all predicates are either Q.gt, Q.lt, Q.ge, Q.le or Q.eq and then makes some calls to `satisfiable` with `use_lra_solver` set to True. The `use_lra_solver` parameter tells satisfiable to use an LRASolver object in conjunction with the dpll2 algorithm.


#### Release Notes

<!-- Write the release notes for this release below between the BEGIN and END
statements. The basic format is a bulleted list with the name of the subpackage
and the release note for this PR. For example:

* solvers
  * Added a new solver for logarithmic equations.

* functions
  * Fixed a bug with log of integers. Formerly, `log(-x)` incorrectly gave `-log(x)`.

* physics.units
  * Corrected a semantical error in the conversion between volt and statvolt which
    reported the volt as being larger than the statvolt.

or if no release note(s) should be included use:

NO ENTRY

See https://github.com/sympy/sympy/wiki/Writing-Release-Notes for more
information on how to write release notes. The bot will check your release
notes automatically to see if they are formatted correctly. -->

<!-- BEGIN RELEASE NOTES -->

* assumptions
  * Linear inequalities and equalities with real variables and rational coefficients are now understood by `ask`. For example,  queries like `ask(x > 1, x> 0)` will no longer return None. However, x needs to be initialized as a real variable like this: `x = symbols("x", real=True)`. 

* logic 
  *  The dpll2 algorithm now has the option to use an LRA theory solver.

<!-- END RELEASE NOTES -->

TiloRC committed:
Move find_feasible code from #25277 to new PR

TiloRC committed:
semi-working implementation of the paper

Still needs to be tested more thoughroughly.
There are probably more bugs. Currently in the
processes of getting rid of bug.

TiloRC committed:
small change

Upabjojr committed:
added debug printer to LRASolver.check

TiloRC committed:
fix -A[i, j]

TiloRC committed:
rename: lra.beta --> lra.assign

TiloRC committed:
check: rename local A to M to avoid confusion with self.A

TiloRC committed:
check: make basic, nonbasic variables

basic and nonbasic are distinct from self.basic and self.nonbasic

TiloRC committed:
implement _debug_internal_state_printer

TiloRC committed:
Changed how pivoting worked

Previously the code was doing pivoting as described
in Ferguson's notes. I moved to a more standard way
of pivoting.

TiloRC committed:
Fixed bug!!!!!!

TiloRC committed:
Clean up some tests

TiloRC committed:
Clean up code

Deleted commented out code.

TiloRC committed:
reorganize variables

self.basic and self.nonbasic variables have been renamed
to self.slack and self.nonslack.

TiloRC committed:
Deleted unused line: j = nonbasic[xj]

TiloRC committed:
check: reorganize basic and nonbasic variables

TiloRC committed:
return variable assignment when SAT

TiloRC committed:
Standardize return type of assert_con to always be tupple

TiloRC committed:
find_feasible: include conflict clause in retured tuple

TiloRC committed:
LRASolver: Add variables parameter to preproccesing

TiloRC committed:
LRASolver.__init__: add assert statement to check A formatting

A should have a negative identity matrix at the end for slack
variables.

TiloRC committed:
remove obsolete code

TiloRC committed:
LRASolver.preprocess: Use dummy variables

TiloRC committed:
LRASolver.preprocess: handle inequalities not in standard form

TiloRC committed:
Fix random testing by setting sparcity to 0

TiloRC committed:
print assignments in debugger

TiloRC committed:
add outline for new method: from_encoded_cnf

TiloRC committed:
improve rand problem generation: make simpler problems

The coefficients of variables in constraints are ints by default
instead of Rationals now.

smichr committed:
pep8

TiloRC committed:
added assert within check():

assignments for x must always satisfy Ax = 0

TiloRC committed:
fix bug with pivot

basic variables should be negative

TiloRC committed:
fix pivot

I think this should work. I don't think I'll change it again.

TiloRC committed:
thousands of random tests all pass

TiloRC committed:
Merge branch 'LRA-solver' of https://github.com/TiloRC/sympy into LRA-solver

TiloRC committed:
progress so far

TiloRC committed:
strict inequalities now work

However, conflict clauses and assignments no longer work.

TiloRC committed:
Test cases seem to be passing

TiloRC committed:
bug fix: _assert_upper/lower never returns sat

This was causing a bug. I don't know why the pseudo code
from paper was doing it that way.

TiloRC committed:
conflict clauses are being generated again

This is mostly to check if everything is working.
I have yet to implement them in a way that works
nicely with SAT solver.

TiloRC committed:
conflict clauses now returned as encoded literals

TiloRC committed:
created docstring for _pivot

TiloRC committed:
conflict clauses are now negative

TiloRC committed:
LRASolver: added docstring and comments to from_encoded_cnf

TiloRC committed:
delete print statements

TiloRC committed:
depreciate functions made obsolete by from_encoded_cnf

TiloRC committed:
write something for check and assert_enc_boundry docstrings

TiloRC committed:
skip doctests because of nondetermism

TiloRC committed:
update and remove tests for depeciated code

TiloRC committed:
move test_random_problem to bottom of testing file

TiloRC committed:
remove old print statement

TiloRC committed:
add test for pivot operation

TiloRC committed:
remove changes from solvers module

all code should be in the logic module now

TiloRC committed:
clean up any changes I missed in Solver module

TiloRC committed:
remove any changes I missed in solvers module

fogot to add inequalities file; only added testing file in last
commit.

TiloRC committed:
remove unused imports and fix spacing

TiloRC committed:
remove costly rref assertion in _pivot

Tests have been added for this in
test_lra_theory.py.

TiloRC committed:
spacing

TiloRC committed:
add returns section to check docstring

TiloRC committed:
Split long lines over multiple lines

TiloRC committed:
check uses True/False instead of SAT and UNSAT

TiloRC committed:
add returns section to assert_enc_boundry docstring

TiloRC committed:
clean up print statements

TiloRC committed:
remove newline at file end

TiloRC committed:
attempt to remove unicode

TiloRC committed:
remove unnesesary set comprehension with list comp

TiloRC committed:
removed unassigned var: Mi, Mj

TiloRC committed:
remove unassigned var: r = random()

TiloRC committed:
fix Unnecessary generator: rewrite as set comprehension

TiloRC committed:
fix: Unnecessary generator - rewrite as a set comprehension

TiloRC committed:
reword check docstring

TiloRC committed:
WIP: start implementing backtracking

TiloRC committed:
LRA Solver: add reset_bounds()

TiloRC committed:
ready LRASolver for SAT integration

- return SAT without having to call Check() when possible
- handle assering ints without an encoding
- fix bug with Q.ne

TiloRC committed:
Start integrating LRASolver w dpll2

TiloRC committed:
add failing test case

sylee957 commented on sympy/logic/algorithms/lra_theory.py:
```diff
@@ -0,0 +1,667 @@
+        if self.run_checks:
+            assert A[:, n-m:] == -eye(m)
+
+        self.boundry_enc = boundry_enc
+        self.boundry_rev_enc = {value: key for key, value in boundry_enc.items()}
+
+        self.A = A # TODO: Row reduce A
+        self.slack = slack_variables
+        self.nonslack = nonslack_variables
+        self.all_var = nonslack_variables + slack_variables
+        self.col_index = {v: i for i, v in enumerate(self.all_var)}
+
+        # if previously we knew this was sat
+        # if no assertions about slack variables have been made since
+        # and changing bounds on nonslack hasn't lead to unsat
+        self.is_sat = True
+        self.slack_set = set(slack_variables) # used for checking if var is slack
+
+        self.lower = {}
+        self.upper = {}
+        self.low_origin = {}
+        self.up_origin = {}
+        self.assign = {}
+
+        # always one of: (True, Assignment), (False, ConflictClause), None
+        self.result = None
+
+        for var in self.all_var:
+            self.lower[var] = (-float("inf"), 0)
+            self.low_origin[var] = False
+            self.upper[var] = (float("inf"), 0)
+            self.up_origin[var] = False
+            self.assign[var] = (0, 0)
+
+        # Backtracking Variables
+        # stack contains elements of form (var, lower, upper)
+        # where lower or upper may be None
+        self.stack_bounds = []
+        # only assignment from the last succesful check needs to be stored
+        self.last_safe_assignment = self.assign.copy()
```
Please try to annotate the mutable fields with types.

sylee957 commented on sympy/logic/algorithms/lra_theory.py:
```diff
@@ -0,0 +1,667 @@
+        return hash((self.var, self.bound, self.strict, self.upper, self.equality))
+
+
+class LRASolver():
```
I'm not sure if this is intended for inheritance or not.
If you are not sure, or experimental, I'd just try to decorate `@typing.final` to keep the users out of extending the objects.

TiloRC replied:
It's not. This specific part of the code is actually from https://github.com/sympy/sympy/pull/25363.

asmeurer replied:
I don't see what the point is of explicitly disabling subclassing like that. Especially for a class that's just for internal use anyway.

sylee957 replied:
The point is we need to discuss about ‘design’ issue before making anything public.

If we are still experimenting, or avoid design reviews, I just want to hide that as much from users.

moorepants replied:
That's fine, but you should probably open a SYMPEP to outline any things you want the whole SymPy community to follow. Once we agree on a SYMPEP, then we can enforce/encourage the policies developed there.

TiloRC replied:
Maybe I should a leading underscore so it's _LRASolver?

sylee957 replied:
That may work

TiloRC committed:
fix bugs introduced by changes from the integration PR

TiloRC committed:
30x speed up in from_enc_cnf

TiloRC commented on sympy/logic/algorithms/lra_theory.py:
```diff
@@ -0,0 +1,667 @@
+        Example
+        =======
+
+        >>> from sympy.matrices.dense import Matrix
+        >>> from sympy.logic.algorithms.lra_theory import LRASolver
+        >>> from sympy import var
+        >>> Matrix(3, 3, var('a:i'))
+        Matrix([
+        [a, b, c],
+        [d, e, f],
+        [g, h, i]])
+
+        This matrix is equivalent to:
+        0 = a*x + b*y + c*z
+        0 = d*x + e*y + f*z
+        0 = g*x + h*y + i*z
+
+        >>> LRASolver._pivot(_, 1, 0)
+        Matrix([
+        [ 0, -a*e/d + b, -a*f/d + c],
+        [-1,       -e/d,       -f/d],
+        [ 0,  h - e*g/d,  i - f*g/d]])
+
+        We rearage equation 1 in terms of variable 0 (x)
+        and substitute to remove x from the other equations.
+
+        0 = 0 + (-a*e/d + b)*y + (-a*f/d + c)*z
+        0 = -x + (-e/d)*y + (-f/d)*z
+        0 = 0 + (h - e*g/d)*y + (i - f*g/d)*z
+        """
+        _, _, Mij = M[i, :], M[:, j], M[i, j]
+        if Mij == 0:
+            raise ZeroDivisionError("Tried to pivot about zero-valued entry.")
+        A = M.copy()
+        A[i, :] = -A[i, :]/Mij
+        for row in range(M.shape[0]):
+            if row != i:
+                A[row, :] = A[row, :] + A[row, j] * A[i, :]
+
+        return A
```
@oscarbenjamin could you give any advice on how to speed up this pivot function? Currently it's the bottleneck in the code.

TiloRC replied:
<img width="589" alt="Screen Shot 2023-08-18 at 7 06 29 PM" src="https://github.com/sympy/sympy/assets/67246777/81b35817-2eb9-4c98-89b8-2722cf732cb8">

sylee957 replied:
How did you reach to the example?
I would take a look at if there is a regression about assignment because of the introduction of `RepMatrix`
https://github.com/sympy/sympy/pull/21626

TiloRC replied:
> How did you reach to the example?
I made 20 largish random problems using the code in test_lra_theory.py and used the LRASolver to solve them. 

> I would take a look at if there is a regression about assignment because of the introduction of RepMatrix
https://github.com/sympy/sympy/pull/21626

I'm a bit confused about what this means. Regression as in a test for some fixed bug? What do you mean by "assignment"?

sylee957 replied:
Assignment here means `M[i, j] = ...`

sylee957 replied:
> I made 20 largish random problems using the code in test_lra_theory.py and used the LRASolver to solve them.

Can you refer to the specific test case?

TiloRC replied:
```
problems = [make_random_problem(num_variables=8, num_constraints=16, rational=False, disable_strict=False,
                                  disable_nonstrict=False, disable_equality=False) for _ in range(20)]

problems = [prob_to_enc(cons) for cons in problems]

from pyinstrument import Profiler

profiler = Profiler()
profiler.start()
for enc in problems:
    lra, x_subs, s_subs = LRASolver.from_encoded_cnf(enc)
    assert_bounds(lra, enc)
    lra.check()

profiler.stop()
#profiler.open_in_browser()
```

The screenshot is from lra.check().

Since `num_variables=8, num_constraints=16`, the matrices involved would be around 16 by 24. Basically with these settings each random problem has 16 equalities, strict or non-strict inequalities and each of those constraints would have 8 variables. And the numbers in the matrices would be small integers at first.

sylee957 replied:
Anyway, I can confirm that there is no much difference about forcing the Matrix to `EXRAW`.
It means that we may need hard time around, to optimize the algorithm itself.

oscarbenjamin replied:
`Matrix.copyin_matrix` can be made more efficient by adding a corresponding DomainMatrix method and using that.

TiloRC committed:
added profiling code to testing

faze-geek commented on sympy/logic/algorithms/lra_theory.py:
```diff
@@ -0,0 +1,667 @@
+
+        Conceptually, M represents a system of equations and pivoting
+        can be thought of as rearranging equation i to be in terms of
+        variable j and then substiututing in the rest of the equations
```
"substituting"

TiloRC committed:
Merge branch 'LRA-solver' into lra-dpll2

TiloRC committed:
Add aditional assertions for testing and debugging

TiloRC committed:
spelling: substiututing -> substituting

TiloRC committed:
Ensure input is rational

TiloRC committed:
dramatic speed up in from_encoded_cnf

Makes 30% faster in test random problems. However,
biggest speedups were seen in problems that don't
involve any constraints.

Speeds assumptions/tests by 50%. As a result, LRASolver
doesn't slow down tests lacking constraints.

TiloRC commented:
Something that's not implemented and will be a pain in the ass to implement:
```
satisfiable( ~(x > 1) & ~(x < 1)) # current code returns SAT as it should because x may be imaginary 
satisfiable( ~(x > 1) & ~(x < 1) & Q.real(x)) # current code returns SAT, but it shouldn't 
```

TiloRC committed:
Handle Q.pos, neg, pos_inf, neg_inf, zero

TiloRC committed:
bug fixes and Q.gt(3, 2) handling

Fixed multiple bugs:

- Matrix expresions being added to constants

- Handling for empty predicate like Q.positive
(instead of Q.positive(x))

Also, binrel with only constants such as Q.gt(3, 2)
are now handled.

TiloRC committed:
handle constraints containing nan – ignore them

TiloRC committed:
Introduce ExtendedRational class to allow for infinity

TiloRC committed:
Infinity should be handled. Further testing needed.

TiloRC committed:
start implementing _remove_unneeded_variables

TiloRC committed:
rename enc_boundry and assert_enc_boundry for clarity

TiloRC committed:
Merge branch 'z3' into lra-dpll2

TiloRC committed:
Fix bugs with dpll2 / lra integration

TiloRC committed:
add test comparing z3 with lra_dpll2

TiloRC committed:
fix theory conflict clause learning and improve tests

TiloRC committed:
refactor: Use VariableLRA to keep track of upper/lower

code is a lot simpler

TiloRC committed:
raise exceptions for various unhandled input

TiloRC committed:
cleaned up code and removed x and s_subs from from_enc... return

TiloRC committed:
added xfailing test for propper negation handling

TiloRC committed:
remove changes to satask

I decided to decrease the scope of what I was planning to do.
Instead of integrating with satask (which is a longterm
goal), I will make a temporary lra_satask function
which handles inequalities if all variables are finite.

TiloRC committed:
Merge branch 'z3' into lra-dpll2

TiloRC committed:
code quality

TiloRC committed:
code quality - list comp

TiloRC committed:
deleted old commented out code

TiloRC committed:
small fix

TiloRC committed:
Start implementing integration with ask

TiloRC committed:
added negation handling

TiloRC committed:
unequality splitting enabled

TiloRC committed:
skip if z3 is not installed

TiloRC committed:
z3_sat: raise error if z3 is not installed

TiloRC committed:
from_enc_cnf: test for matrix expr with .kind

TiloRC committed:
from_encoded_cnf: improved docstring

TiloRC committed:
ask: added tests involving inequalities

TiloRC commented on sympy/logic/tests/test_lra_theory.py:
```diff
@@ -0,0 +1,482 @@
+    profiler = Profiler()
+    profiler.start()
+    for enc in problems:
+        lra, x_subs, s_subs = LRASolver.from_encoded_cnf(enc)
```
I changed the return type of LRASolver.from_encoded_cnf at some point so that may have been the issue. A lot of stuff was broken a week ago.

TiloRC committed:
many small improvements to code quality and docstrings

TiloRC committed:
move helper methods and classes to bottom of file

oscarbenjamin commented on sympy/logic/algorithms/lra_theory.py:
```diff
@@ -0,0 +1,840 @@
+        return None
+
+
+class Boundry:
```
Boundary

oscarbenjamin commented on sympy/logic/algorithms/lra_theory.py:
```diff
@@ -0,0 +1,840 @@
+
+class LRASolver():
+    """
+    Linear Arithmatic Solver for DPLL(T) implemented with algorithm based on
```
Arithmetic

oscarbenjamin commented:
I don't know enough the background here to review this in detail but it looks very nice.

TiloRC committed:
handle more predicate: Q.pos, Q.neg, and more

TiloRC committed:
fixed an edge case

TiloRC committed:
spelling: Boundry --> Boundary

smichr commented on sympy/logic/algorithms/lra_theory.py:
```diff
@@ -0,0 +1,774 @@
+        [-1,       -e/d,       -f/d],
+        [ 0,  h - e*g/d,  i - f*g/d]])
+
+        We rearage equation 1 in terms of variable 0 (x)
```
```suggestion
        We rearrange equation 1 in terms of variable 0 (x)
```

smichr commented on sympy/logic/algorithms/lra_theory.py:
```diff
@@ -0,0 +1,774 @@
+
+                raise ValueError(f"Unhandled Predicate: {prop}")
+
+
+
```
```suggestion
```

smichr commented on sympy/logic/algorithms/lra_theory.py:
```diff
@@ -0,0 +1,774 @@
+        else:
+            encoded_cnf_items = encoded_cnf.encoding.items()
+
+
```
```suggestion
```

asmeurer commented on sympy/logic/algorithms/lra_theory.py:
```diff
@@ -0,0 +1,774 @@
+    """
+    def __init__(self, var, const, upper, equality, strict=None):
+        if not equality in [True, False]:
+            assert equality in [True, False]
```
You should just use ValueError here.

asmeurer commented on sympy/logic/algorithms/lra_theory.py:
```diff
@@ -0,0 +1,774 @@
+        self.is_sat = True
+        self.result = None  # always one of: (True, assignment), (False, conflict clause), None
+
+    @staticmethod
```
This should be a classmethod.

sylee957 replied:
I have thought similar that staticmethod can always be redundant in Python before, especially for everything was duck typed.

However, now I just think that there is also no use of classmethod if the method does not depend on the parameter cls.

Formally, it should be possible to substitute  staticmethod, with ‘def function’, as well as trivial class methods or trivial bound methods that does not depend on the class, object.

So I stopped thinking that that coding was a problem, especially after type checkers or linters are established.

asmeurer replied:
The reason I say it should be a class method is because it's an alternative constructor, which is the typical use-case of class methods in Python. However, in this case, it returns an instance of the class and also something else, so maybe it isn't really a constructor.

asmeurer replied:
Also this would depend on cls. The function uses `LRASolver`, which would be `cls` if this were a class method.

sylee957 replied:
Okay, I thought it was pointing out to `pivot` though

asmeurer commented:
@jksuom I don't know how familiar you are with this sort of thing, but if you wanted to take a look here and review this it would be helpful.

TiloRC commented on sympy/logic/algorithms/lra_theory.py:
```diff
@@ -0,0 +1,774 @@
+
+
+            assert prop.function in ALLOWED_PRED
+            if prop.lhs.kind == MatrixKind(NumberKind) or prop.rhs.kind == MatrixKind(NumberKind):
+                raise ValueError(f"{prop} contains matrix variables")
```
@oscarbenjamin what do you think should happen here if .kind returns UndefinedKind?

Would something that returned UndefinedKind be something that this LRASolver might handle? Are there any examples that return UndefinedKind that I could use in the tests?

oscarbenjamin replied:
Currently we have:
```python
In [1]: cos(x).kind
Out[1]: UndefinedKind
```
Ideally that would be changed but most things are in reality `NumberKind` (regardless of what `.kind` gives) and most of the time `UndefinedKind` really means `NumberKind` just because the system is incomplete.

For a test I would just make a new Expr subclass.

TiloRC committed:
Merge branch 'master' into lra-dpll2

TiloRC committed:
LRASolver: depreciate x_subs

TiloRC committed:
Added explanation to top of file

TiloRC committed:
added more comments and expanded on some docstrings

TiloRC committed:
deleted uneeded test

TiloRC committed:
from_encoded_cnf: add comments and increase readability

TiloRC committed:
move debugging functions out of the way

TiloRC committed:
added a few comments

TiloRC committed:
UnhandledNumber --> UnhandledInput

Wanted a more general exception.

TiloRC committed:
raise UnhandledInput when nonlinearity is present

TiloRC committed:
handle matrix expr in lra_satask instead of LRASolver

TiloRC committed:
lra_satask: raise inconsistent assumptions exception

TiloRC committed:
expand allowed pred to include ignorable pred

TiloRC committed:
added more relational ask tests and fixed bug

TiloRC committed:
fixed some preprocessing bugs with edge cases

asmeurer commented:
I get a lot of test failures when I test this locally

```
FAILED sympy/assumptions/tests/test_rel_queries.py::test_lra_satask - assert None is False
FAILED sympy/assumptions/tests/test_rel_queries.py::test_rel_queries - assert None is False
FAILED sympy/assumptions/tests/test_rel_queries.py::test_all_pred - assert None is True
FAILED sympy/assumptions/tests/test_rel_queries.py::test_number_line_properties - assert None is True
FAILED sympy/assumptions/tests/test_rel_queries.py::test_equality - assert None is True
```

and even the example in the OP doesn't work

```
>>> x, y, z = symbols("x y z", real=True)
>>> print(lra_satask(x > z, (x > y) & (y > z)))
None
```

asmeurer commented on sympy/logic/inference.py:
```diff
@@ -77,6 +77,12 @@ def satisfiable(expr, algorithm=None, all_models=False, minimal=False):
     """
+    if use_lra_theory:
+        if algorithm is not None:
+            assert algorithm == "dpll2"
```
This should use `ValueError` not `AssertionError`. Assertions should not be used to check user inputs.

asmeurer replied:
And elsewhere too, an AssertionError should only ever occur if there is a bug in the code.

asmeurer commented:
FWIW, I think the biggest limitation here that we'll want to address in the future is the requirement that the variables be set as real using the old assumptions. It's presently a bit of a gotcha that something like `ask(x > 0, x > 1)` works if `x` is set as `real=True` but gives None if it is defined without any assumptions.

asmeurer commented:
We currently have this behavior

```py
>>> x = symbols('x', positive=True)
>>> ask(x < 0, x < -1)
False
```

But this should give an error. I think we can fix this by making `positive=True` and `negative=True` (and so on) automatically insert `x > 0`, `x < 0`, etc. into the assumptions. That way, the above would be equivalent to 

```py
>>> x = symbols('x', real=True)
>>> ask(x < 0, (x < -1) & (x > 0))
```

which does properly raise an inconsistent assumptions error.

asmeurer commented on sympy/logic/algorithms/lra_theory.py:
```diff
@@ -0,0 +1,951 @@
+
+    First a preprocessing step should be done on f. During preprocessing,
+    f should be checked for any predicates such as `Q.prime` that can't be
+    handled. Also disequalities like `~Eq(y, 1)` should be split.
```
A better word is "unequality". That's also what the SymPy Ne class is called.

```
>>> str(Ne)
"<class 'sympy.core.relational.Unequality'>"
```

asmeurer commented on sympy/logic/algorithms/lra_theory.py:
```diff
@@ -0,0 +1,951 @@
+    to a SAT solver. The SAT solver will give us a satisfying
+    assignment such as:
+
+    Eq(1, 2): False
+    Eq(y,  1): True
+    (y < 1) : True
+    (y > 1) : True
+    (x = 0): True
+    (x < 0): True
+    (x > 0): True
```
```suggestion
    (1 = 2): False
    (y = 1): True
    (y < 1): True
    (y > 1): True
    (x = 0): True
    (x < 0): True
    (x > 0): True
```

asmeurer commented on sympy/logic/algorithms/lra_theory.py:
```diff
@@ -0,0 +1,951 @@
+    boolean formula was unsatisfiable.
+
+
+The paper this implementation comes from includes a
```
```suggestion
This implementation is based on [1]_, which includes a
```

asmeurer commented on sympy/logic/algorithms/lra_theory.py:
```diff
@@ -0,0 +1,951 @@
+detailed explanation of the algorithm and pseudocode
+for the most important functions.
+
+The paper also explains how backtracking and theory propagation
+could be implemented to speed up the current implementation.
```
```suggestion
[1]_ also explains how backtracking and theory propagation
could be implemented to speed up the current implementation, 
but these are not currently implemented.
```

asmeurer commented on sympy/logic/algorithms/lra_theory.py:
```diff
@@ -0,0 +1,951 @@
+    Raised while creating an LRASolver if non-linearity
+    or non-rational numbers are present.
+    """
+    pass
```
`pass` is not needed here 

```suggestion

```

asmeurer commented on sympy/logic/algorithms/lra_theory.py:
```diff
@@ -0,0 +1,951 @@
+
+    tableau = Matrix([[corner, r1], [c1, A]])
+    pprint(tableau)
+    sys.stderr.write("\n")
```
It's better to use `print(file=sys.stderr)`. Also you can use `pretty()` instead of `pprint()` to get the string form so it is also printed to stderr.

asmeurer replied:
Actually, why not just use the `debug()` function?

asmeurer commented on sympy/logic/tests/test_lra_theory.py:
```diff
@@ -0,0 +1,505 @@
+            # check that conflict clause is probably minimal
+            for subset in itertools.combinations(conflict, len(conflict)-1):
+                assert check_if_satisfiable_with_z3(subset) is True
+    # print("check", check_time)
+    # print("from_encoded", from_encoded_time)
+    # print("assert", assert_time)
+    # print("total", check_time+from_encoded_time+assert_time)
+    # print("sat count", feasible_count)
```
```suggestion
```

asmeurer replied:
Also remove all the timing stuff above that this is coming from.

If we want to track the performance here, we should add benchmarks to the benchmarks suite https://github.com/sympy/sympy_benchmarks.

asmeurer commented on sympy/logic/tests/test_lra_theory.py:
```diff
@@ -0,0 +1,505 @@
+    # Version1 is not compatible with version2.
+
+    # note that SymPy assumes variables are complex by default
+    # thus Not(x > 0) implies (x <= 0) OR (x has imaginary component)
```
It's also not actually currently the case that `I > 0` is False. That's what I'd propose for it to do, at least internally in the assumptions, but presently it gives an error.

asmeurer commented on sympy/logic/tests/test_lra_theory.py:
```diff
@@ -0,0 +1,505 @@
+from sympy.testing.pytest import raises, XFAIL, skip
+import time
+
+def make_random_problem(num_variables=2, num_constraints=2, sparsity=.1, rational=True,
```
This doesn't need to be done in this pull request, but I'd like to investigate replacing  this (or at least adding in addition to this) with hypothesis strategies.

asmeurer commented on sympy/logic/tests/test_inference.py:
```diff
@@ -323,10 +331,63 @@ def test_z3():
+            sat_count += 1
+
+    # print("lra_dpll2_time", lra_dpll2_time)
+    # print("z3_time", z3_time)
```
Same here, let's delete all the timing stuff from the test.

asmeurer commented on sympy/assumptions/ask.py:
```diff
@@ -497,6 +498,10 @@ def ask(proposition, assumptions=True, context=global_assumptions):
+    if res is not None:
+        return res
+
+    res = lra_satask(proposition, assumptions=assumptions, context=context)
```
I would suggest refactoring this so that lra_satask raises UnhandledInput exceptions, and the logic here catches those exceptions and returns None. That way, you can call lra_satask directly if you want to be clear about whether something is returning None because it is actually undetermined or just because the input currently can't be handled.

TiloRC committed:
Apply suggestions from code review

Various small changes suggested such as spelling and spacing and deleting commented out code

Co-authored-by: Aaron Meurer <asmeurer@gmail.com>
Co-authored-by: Christopher Smith <smichr@gmail.com>

TiloRC committed:
delete all print statements

TiloRC committed:
infernece: Raise value error instead of assertion error

Raising more infomrative error message about how dpll2
is the only currently supported lra theory algorithm.

TiloRC committed:
add pycosat to CI optional dependencies build

TiloRC committed:
catch UnhandledInput in ask instead of lra_satask

TiloRC committed:
use old assumptions

TiloRC committed:
disequality --> unequality

asmeurer commented on sympy/logic/tests/test_inference.py:
```diff
@@ -388,6 +388,3 @@ def make_random_cnf(num_clauses=5, num_constraints=10, num_var=2):
             sat_count += 1
-
-    # print("lra_dpll2_time", lra_dpll2_time)
-    # print("z3_time", z3_time)
```
Also remove the timing stuff from the test. If we want to keep track of timings we should do that in the benchmarking suite.

TiloRC commented:
@asmeurer been a bit busy with college so I haven't worked on this in a few days much. How close do you think this PR is to merging? 

As it stands there are only a few things I need to change:
- seems like I missed a few commented out pieces of code that should be deleted 
- either get rid of the debug function or address comments about it and possibly improve it in other ways
- change some static methods to class methods

asmeurer commented:
I think the only thing here is removing the extra print statements and debug code. I wouldn't worry about the staticmethod vs. classmethod thing.

TiloRC commented:
I had an insight in the best way to make it so you don't have to explicitly specify variables are real.

Currently, the reason why you have to specify this is because subtleties of negating inequalities when variables are allowed to be complex; `~Q.gt(x, y)` is not equivalent to `Q.le(x,y)` like it is when x and y must be reals.

My idea is to do things in a similar way to how I implemented the z3 wrapper. How that works is that a formula like `(x > 0) & (x < 0)` becomes `b1 & b2 & (b1 $\implies$ (x > 0)) & (b2 $\implies$ (x < 0))`.

This approach has the advantage that we can keep the implementation of the LRA Solver as close to the paper it was based on as possible. Also, this should have a speed advantage; the conflict clauses produced by the LRA solver will all be over the reals which will allow it to take advantage of the `~Q.gt(x, y)`  if an only if `Q.le(x,y)` property.  Also this shouldn't be too bad to implement.

However, it might be better to do this in a follow up pull request.

TiloRC commented on sympy/assumptions/lra_satask.py:
```diff
@@ -0,0 +1,285 @@
+    return check_satisfiability(props, _props, assumptions)
+
+
+WHITE_LIST = ALLOWED_PRED | BENIGN_PRED | {Q.positive, Q.negative, Q.zero, Q.nonzero, Q.nonpositive, Q.nonnegative,
```
I should add some comments explaining what WHITE_LIST and the other variables here are

TiloRC replied:
something like:

# Some predicates such as Q.prime can't be handled by lra_satask.
# For example, (x > 0) & (x < 1) & Q.prime(x) is unsat but lra_satask would think it was sat.
# WHITE_LIST is a list of predicates that can always be handled.

TiloRC commented on sympy/assumptions/lra_satask.py:
```diff
@@ -0,0 +1,285 @@
+    sat_true.add_from_cnf(prop)
+    sat_false.add_from_cnf(_prop)
+
+    all_pred, all_exprs = get_all_pred_and_expr(sat_true)
```
maybe `enc_cnf_to_pred_and_expr` would be a better name

TiloRC commented on sympy/assumptions/lra_satask.py:
```diff
@@ -0,0 +1,285 @@
+
+    return all_pred, all_exprs
+
+def extract_old_assumptions(all_exprs):
```
maybe `extract_pred_from_old_assum` would be a better name.

TiloRC commented on sympy/assumptions/lra_satask.py:
```diff
@@ -0,0 +1,285 @@
+
+    all_pred, all_exprs = get_all_pred_and_expr(sat_true)
+
+    for pred in all_pred:
+        if pred.function not in WHITE_LIST and pred.function != Q.ne:
+            raise UnhandledInput(f"LRASolver: {pred} is an unhandled predicate")
+    for expr in all_exprs:
+        if expr.kind == MatrixKind(NumberKind):
+            raise UnhandledInput(f"LRASolver: {expr} is of MatrixKind")
+        if expr == S.NaN:
+            raise UnhandledInput("LRASolver: nan")
```
put this into a helper function—maybe `check_for_unhandled_input`

asmeurer commented:
Yes, let's implement ideas like that in a followup PR. I'd like to get this PR merged ASAP.

TiloRC commented on sympy/assumptions/lra_satask.py:
```diff
@@ -0,0 +1,285 @@
+        if expr == S.NaN:
+            raise UnhandledInput("LRASolver: nan")
+
+    # add any old assumptions and raise
+    # exceptions if there are any unhandled assumptions
```
Improve comments:
```
# convert old assumptions into predicates and add them to sat_true and sat_false
# also check for unhandled inputs
```

We can reuse the check_for_unhandled_input function here.

asmeurer commented:
> Currently, the reason why you have to specify this is because subtleties of negating inequalities when variables are allowed to be complex; ~Q.gt(x, y) is not equivalent to Q.le(x,y) like it is when x and y must be reals.
>
> My idea is to do things in a similar way to how I implemented the z3 wrapper. How that works is that a formula like (x > 0) & (x < 0) becomes b1 & b2 & (b1 $\implies$ (x > 0)) & (b2 $\implies$ (x < 0)).

There might be more work required here because SymPy doesn't currently actually have `(x > y) == False` for non-real x and y. So there might be places that need to be adapted for that. Definitely the core inequality classes won't work (they raise an exception if they know the inputs aren't real), but it could be the case for the Q.le classes as well.

Although perhaps we could use a similar idea to make setting things as real with the new assumptions work. If `x > y` is an inequality in the input, add the fact `(x > y) >> (Q.real(x) & Q.real(y))`. If `Q.real(x)` and `Q.real(y)` are provided, this will effectively be ignored by the solver. However, if they are not provided, and the LRA solver produces a model where `x > y` is true, then `ask` will know that this is not possible if `Q.real(x)` can be False and so it will return None.

TiloRC committed:
lra_satask: small improvements to code quality

Renamed a function and added a comment explaning WHITE_LIST.

TiloRC committed:
delete tests for alternative negation handling

Seems like it would be better to discuss alternatives
ways to handle negation in an issue rather than have these
Xfailing tests which could be a bit confusing.

TiloRC committed:
added comments explaining ALLOWED_PRED, BENIGN_PRED, HANDLE_NEGATION

TiloRC commented:
I think it was a mistake to allow predicates such as Q.complex, Q.transcendental, .. etc. 

None of these predicates are actually safe to ignore. ~Q.complex(x) for example contradicts x > 0.

TiloRC committed:
remove benign pred

None of of the benign pred were really benign
when you consider that they could be negated.

oscarbenjamin commented:
> ~Q.complex(x) for example contradicts x > 0.

What is wrong with that?

Note that the reals are a subset of the complexes so real numbers are also complex:
```python
In [102]: ask(Q.complex(1))
Out[102]: True
```
If `~Q.complex(x)` means that `x` is not complex then it also means that `x` is not real.

The only object in SymPy for which `x > 0` makes sense that is not complex is `x = oo`:
```python
In [103]: ask(Q.complex(oo))
Out[103]: False
```

TiloRC committed:
remove debug code

TiloRC commented:
> What is wrong with that?

I didn't consider infinite values would be considered non-complex so my original reasoning was bad. Regardless, I think it was still a mistake to allow these "benign" predicates. 

Currently infinite values are not handled and all the variables must be real so ~Q.complex would be in conflict with this assumption. One of the goals of future work would be to handle infinite values.

oscarbenjamin commented:
> ~Q.complex would be in conflict with this assumption.

I don't understand why this matters.

Why in the context of real linear inequalities do you need to care about someone using an assumption like `~Q.complex(x)`?

Any time such an assumption is used real linear inequalities would not be applicable anyway.

asmeurer commented:
Tilo is right here. The reason is that it was just ignoring these predicates entirely in lra_satask. But in the commit prior to 71a96b92333ed37c6e353ea7b6883020e3afd223, you would get things like:

```
>>> x = symbols('x', real=True)
>>> lra_satask((x > 0), ~Q.complex(x) & (x > 1))
True
```

which is wrong (and indeed ask() itself returns False because it never calls lra_satask in this case).

Actually I expect that my solution of adding the `(x > y) >> (Q.real(x) & Q.real(y))` from https://github.com/sympy/sympy/pull/25544#issuecomment-1736278697 would make this work. But let's do that in a follow up PR.

> Any time such an assumption is used real linear inequalities would not be applicable anyway.

It's not that simple. Assumptions in the new assumptions can appear in arbitrary predicates. The ~Q.complex(x) assumption could appear in a way that makes it irrelevant to the actual deduction, or even in a way that makes it never possible to be true.

asmeurer commented:
Regarding the debug code, you should also remove all the stuff that uses `time` (again, we can add benchmarks separately to https://github.com/sympy/sympy_benchmarks if you want). 

Once that happens, I'm +1 to merging this.

TiloRC committed:
remove usage of time

TiloRC committed:
lra random problems: test 50 rather than 300 problems

Testing 300 problems is time consuming and not necesary.
